snowflake_connector.py

```diff
@@ -0,0 +1,445 @@
+        return tuple.__new__(RetVal, (val1, val2))
+
+
+class SnowflakeConnector(BaseConnector):
+
+    def __init__(self):
+
+        # Call the BaseConnectors init first
+        super(SnowflakeConnector, self).__init__()
+
+        self._state = None
+
+        self._account = None
+        self._username = None
+        self._password = None
+
+    def _get_error_msg_from_exception(self, e):
+        error_code = SNOWFLAKE_ERROR_CODE_UNAVAILABLE
+        error_msg = SNOWFLAKE_ERROR_MSG_UNAVAILABLE
+
+        try:
+            if e.args:
+                if len(e.args) > 1:
+                    error_code = e.args[0]
+                    error_msg = e.args[1]
+                    return "Error Code: {0}. Error Message: {1}".format(error_code, error_msg)
+                elif len(e.args) == 1:
+                    error_msg = e.args[0]
+        except Exception:
+            pass
+
+        return "Error Message: {0}".format(error_msg)
+
+    def convert_value(self, value):
+        if isinstance(value, (bytearray, bytes)):
+            return value.decode('utf-8')
+        elif isinstance(value, (datetime.datetime, datetime.timedelta, datetime.date)):
+            return str(value)
+        else:
+            return value
+
+    def _cleanup_row_values(self, row):
+        return {k: self.convert_value(v) for k, v in row.items()}
+
+    def _handle_test_connectivity(self, param):
+        self.save_progress(TEST_CONNECTIVITY_PROGRESS_MSG)
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        try:
+            self._connection = self._handle_create_connection()
+            cursor = self._connection.cursor()
+
+            cursor.execute(SNOWFLAKE_VERSION_QUERY)
+            if cursor:
+                self.save_progress(TEST_CONNECTIVITY_SUCCESS_MSG)
+                return action_result.set_status(phantom.APP_SUCCESS)
+        except Exception as e:
+            self.save_progress(self._get_error_msg_from_exception(e))
+            return action_result.set_status(phantom.APP_ERROR, TEST_CONNECTIVITY_ERROR_MSG)
+
+    def _handle_run_query(self, param):
+        self.save_progress("In action handler for: {0}".format(self.get_action_identifier()))
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        query = param['query']
+        role = param.get('role')
+        warehouse = param.get('warehouse')
+        database = param.get('database')
+        schema = param.get('schema')
+
+        try:
+            self._connection = self._handle_create_connection(role, warehouse, database, schema)
+            cursor = self._connection.cursor(snowflake.connector.DictCursor)
+            cursor.execute(query)
+            returned_rows = cursor.fetchmany(DEFAULT_NUM_ROWS_TO_FETCH)
+
+            for row in returned_rows:
+                action_result.add_data(self._cleanup_row_values(row))
+
+            self.debug_print("returned_rows: {}".format(returned_rows))
```

Comment:
can this contain sensitive info? not sure what the output of printing this is (if it's DB content or just some object descriptor)

Reply:
It is possible that it could contain sensitive info, and is probably not a good idea to debug print here - I'll remove it.

Reply:
Removed.